FIX Better support large or read-only datasets in `decomposition.DictionaryLearning`

jjerphan opened the pull request:
#### Reference Issues/PRs

Follow-up of #23147.
Fix tentative for https://github.com/scikit-learn/scikit-learn/issues/25165

#### What does this implement/fix? Explain your changes.

In some workflows using coordinate descent algorithms:
 - users might provide NumPy arrays with read-only buffers
 - `joblib` might memmap arrays making their buffer read-only.

Yet the implementation of those algorithms need those buffers to be writable.

This introduces a small copy of the slices of the dataset to make them writable in `_sparse_encode` (the `joblib.Parallel`ed function used in `sklearn.decomposition.sparse_encode`) prior to the call to a `Lasso` instance relying on coordinate descent.

Moreover, `cnp.ndarray` is temporarily used instead of memoryviews to allow for a larger support of the variety of NumPy arrays since `const`-qualified memoryviews aren't yet supported. See #25322 for more details in this regards.

jjerphan committed:
FIX Make `cd_fast` use cnp.ndarray to support readonly buffers

Follow-up of #23147.

ogrisel commented:
Thanks for the fix. We also need a non-regression test for this. I don't really understand why the error is triggered in the first place because the data loaded by keras in the reproducer is actually no backed by a readonly buffer according to `x_train.flags`.

Furthermore, I thought we had a common test that checked that we could fit on readonly-buffers using a numpy array generated via `mmap_mode="r"`.

It would be great to really understand what's going on in the reproducer before going forward.

ogrisel commented:
Update: this common estimator check should have tested that all scikit-learn estimators can be fitted on readonly data buffers:

- https://github.com/scikit-learn/scikit-learn/blob/a576bcc22f0a22d0e9db1e529b21d7e1266f96ca/sklearn/utils/estimator_checks.py#L103

lesteve commented:
I have posted a snippet (without keras) to reproduce in https://github.com/scikit-learn/scikit-learn/issues/25165#issuecomment-1348588230

I think the issue is not that the fitted array is a read-only memmap, but that joblib creates read-only memmaps and that call some cython code with these read-only memmaps, and that cython chokes on read-only arrays when you create memoryviews.

Honestly I don't know how to better test this, you would need:
- arrays big enough (>1MB) to trigger joblib read-only memmap creation
- test all the combination of parameters that could end up calling some cython code which does not support read-only array. In this case there is no issue with the default `fit_algorithm` parameter value

thomasjpfan commented:
> arrays big enough (>1MB) to trigger joblib read-only memmap creation

I do not like it, but we can monkeypatch `Parallel` and adjust `max_nbytes` for testing. In general, there has been issues such as https://github.com/scikit-learn/scikit-learn/issues/19608 where having a public API to adjust `max_nbytes` would be useful.

As for refactoring code to using memoryviews, if our tests do not have enough coverage, then I think we should pause accepting this type of refactoring.

thomasjpfan reviewed (commented):
Overall, moving forward, I think it is safer to keep the existing `cnp.ndarray` for fused types until Cython 3.0 is out.

As for this PR, I think we still need to revert:

https://github.com/scikit-learn/scikit-learn/blob/49ff2d955e393005110cf6f6a1740d681c51ef89/sklearn/linear_model/_coordinate_descent.py#L597-L599

jjerphan committed:
MAINT Remove ReadonlyArrayWrapper

jjerphan committed:
TST Add non-regression test for #25165

This test is built based on Loïc's reproducer:
https://github.com/scikit-learn/scikit-learn/issues/25165#issuecomment-1348588230

Co-authored-by: Loïc Esteve <loic.esteve@ymail.com>

jjerphan committed:
DEBUG Use ReadonlyArrayWrapper to support const qualification

Reverting to use cnp.ndarray does not solve the problem.
This is a tentative workaround using ReadonlyArrayWrapper.

This solves the problem, but there is now a segmentation fault.

thomasjpfan reviewed (commented):
I think this PR is changing too much. I am okay with a reverting #23147 and adding a new non-regression test.

jjerphan commented on sklearn/linear_model/_cd_fast.pyx:
```diff
@@ -565,12 +567,13 @@ def sparse_enet_coordinate_descent(
-    cnp.ndarray[floating, ndim=2, mode='c'] Q,
-    cnp.ndarray[floating, ndim=1, mode='c'] q,
-    cnp.ndarray[floating, ndim=1] y,
+    floating[:, ::1] Q,
+    floating[::1] q,
+    floating[:] y,
```
I've tried both approaches and both do not to prevent the `ValueError` from occurring. I need to take some time to understand why.

jjerphan commented:
It looks like there's something weird at play between NumPy and memory view creation as using `cnp.ndarray` does not resolve the problem for read-only arrays.

The `ValueError` comes from NumPy's `array_getbuffer` (the `ValueError` message is the concatenation of `"buffer source"` (see [here](https://github.com/numpy/numpy/blob/196705a2efffe4d0cc24c92c009ce57237a80cb7/numpy/core/src/multiarray/buffer.c#L774)) and `"array is read-only"` (see [there](https://github.com/numpy/numpy/blob/196705a2efffe4d0cc24c92c009ce57237a80cb7/numpy/core/src/multiarray/arrayobject.c#L635))).

`array_getbuffer` is called (indirectly via Cython macro) via  [`View.MemoryView.memoryview.__cinit__`](https://github.com/cython/cython/blob/0.29.x/Cython/Utility/MemoryView.pyx#L350) I think when memoryviews are created.

Is this a bug of Cython?

thomasjpfan commented:
Just in case it gets lost, I responded to the out of branch commit: https://github.com/scikit-learn/scikit-learn/commit/6c30a489b2d9c338fd060e22c265e9b3305fc92f#r94961757

jjerphan committed:
MAINT Simplify and make copy of chunks of w

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
TST Make test_cd_work_on_joblib_memmapped_data faster to run

jjerphan committed:
fixup! MAINT Simplify and make copy of chunks of w

jjerphan committed:
[scipy-dev] Trigger CI

jjerphan committed:
[scipy-dev] TST cnp.ndarray all the way!

jjerphan committed:
Merge branch 'main' into fix/make-cd_fast-use-nd_array-for-read-only-support

jjerphan committed:
CI Trigger CI

jjerphan committed:
Merge branch 'main' into fix/make-cd_fast-use-nd_array-for-read-only-support

Vincent-Maladiere commented on sklearn/linear_model/_cd_fast.pyx:
```diff
@@ -273,16 +273,17 @@ def enet_coordinate_descent(
     return np.asarray(w), gap, tol, n_iter + 1
 
 
+# TODO: use const fused typed memoryview where possible when Cython 0.29.33 is used.
```
Does this TODO take effect now since const fused typed memoryviews are available?

jjerphan replied:
Not yet, but once https://github.com/scikit-learn/scikit-learn/pull/25342#issuecomment-1377262962 is merged.

ogrisel replied:
#25342 is now merged. Shall we try that now or is it safer not to experiment with const typed memory views in a bugfix (1.2.1) and instead do that only in `main` (for 1.3.0).

jjerphan replied:
Provided that Cython has good coverage and test cases for `const`-qualified memoryviews, I think it is safe and the best to use them to resolve such issues (as it is conceptually the best solution). Yet it do not think we need to use them for this fix with respect to the next bug fix release (I do not want to postpone this release).

Note that using such construct might not be sufficient to fix the problem tackled by this PR (for the reason given in the TODO).

I will start investigating `const`-qualified fused typed memoryviews tomorrow for this fix and for other long standing issues.

thomasjpfan reviewed (commented):
I think this is a big enough fix to add to the 1.2.1 changelog.

jjerphan committed:
Merge remote-tracking branch 'upstream/main' into fix/make-cd_fast-use-nd_array-for-read-only-support

jjerphan committed:
DOC Document workaround

jjerphan committed:
DOC Add a whats_new entry for 1.2.1

jjerphan commented:
I am not entirely satisfied with this fix: I think the root of the problem might better addressed closer to the various coordinate descent algorithms. Yet, I think performing this fix would better come after removing some complexity. Removing this complexity necessitates #25322 to be fixed.

jjerphan committed:
Merge branch 'main' into fix/make-cd_fast-use-nd_array-for-read-only-support

ogrisel reviewed (approved):
LGTM as it is. I think I am fine with experimenting with const typed memory views in a follow-up PR but keep this bugfix as it is for 1.2.1.

lesteve commented:
LGTM, merging this one!
